### PR TITLE
RFC: USB: Use maximum-speed DTS parameters

### LINKED ISSUE #11000: USB 2.0 high-speed support in Zephyr
To the best of my knowledge, Zephyr supports at least one high-speed USB controller, the one in the STM32F723 SoC. The PR #10977 adds the USB driver for the SAM E70 SoC, which also support high-speed USB.

I have done some basic testing of the USB device stack in high-speed mode. So far it seems most of the issues are on the descriptor side. The control and interrupt endpoints work fine, but the bulk endpoints do not always work, as we usually define their size to be 64 bytes, while they should be 512 bytes in high-speed mode. This leads to issues like this one with the Linux kernel:

```
usb 2-2.2: new high-speed USB device number 9 using xhci_hcd
usb 2-2.2: config 1 interface 0 altsetting 0 bulk endpoint 0x81 has invalid maxpacket 64
usb 2-2.2: config 1 interface 0 altsetting 0 bulk endpoint 0x2 has invalid maxpacket 64
usb 2-2.2: New USB device found, idVendor=2fe3, idProduct=0100, bcdDevice= 0.11
usb 2-2.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
usb 2-2.2: Product: Zephyr MSC sample
usb 2-2.2: Manufacturer: ZEPHYR
usb 2-2.2: SerialNumber: 0.01
```

In practice it works or not depending on the host, I got it working on a host with an EHCI controller, but I experienced data loss with an XHCI controller. On the other hand, after increasing the endpoints size to 512 bytes (by changing the limits in the Kconfig code), all work as expected.

I believe that the way to fix that is to fix the descriptors just after the USB reset, when the connection speed is known. The bulk descriptors should be reduced to 64 bytes. That is probably also the good place to fix the `bInterval` values, and provide a Device Qualifier Descriptor.

Now I am not fully sure how to re-architecture the code to be able to do that. I am therefore opening this issue in the hope to get some guidelines. I guess @finikorg and @jfischer-phytec-iot could help there.

### PR DESCRIPTION
**Pull request by finikorg**
Add better HS support

Fixes #11000

### CODE REVIEW

**Comment by aurel32** [comment]
> At the moment maximum-speed is a string making it impossible to use
> with preprocessor macros.

The initial version of the PR didn't use strings. It has been switched to match the linux bindings. See:
https://github.com/zephyrproject-rtos/zephyr/pull/9062#pullrequestreview-146441447

**Comment by carlescufi** [comment]
@aurel32 understood, thanks for that.
@erwango can you also weigh in here?

@finikorg @jfischer-phytec-iot I think it might be better to stay with strings given @galak's original comment

File: include/dt-bindings/usb/usb.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_USB_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_USB_H_
+
+#define USB_MAXIMUM_SPEED_LOW_SPEED	0
```
Comment (carlescufi) [inline include/dt-bindings/usb/usb.h]: I would move those to `usb_dc.h` or similar and then have some code that translates the string into a variable containing one of the values. I would do this directly in PR #14977

**Review by carlescufi — DISMISSED** [review]

File: include/dt-bindings/usb/usb.h (hunk 1 above)
Comment (finikorg) [reply include/dt-bindings/usb/usb.h]: we need this kind of stuff 
b441bba68a64a3ba8a6f389b4845b8878d632921

Comment (finikorg) [reply include/dt-bindings/usb/usb.h]: And especially here 6098067af34951b0ef7bf2ab1dae2721213f49a1. This is just the example, should be changed to every place we define endpoints.

**Review by galak — DISMISSED** [review]
The DTS proper maximum-speed should be a string, it should NOT be a int.  This matches dts binding specs from Linux for this property.

**Comment by finikorg** [comment]
> The DTS proper maximum-speed should be a string, it should NOT be a int. This matches dts binding specs from Linux for this property.

How can we use strings for preprocessor macros?

**Comment by finikorg** [comment]
Maybe our dts parser script can create out of strings like `DT_USB_MAXIMUM_SPEED = "high-speed"` something like
```
#define DT_USB_MAXIMUM_SPEED_HIGH_SPEED 1
```

File: include/usb/usb_device.h [hunk 2]
```diff
@@ -74,6 +75,17 @@ extern "C" {
 
 #define MAX_PACKET_SIZE0    64        /**< maximum packet size for EP 0 */
 
+#if defined(DT_USBHS_MAXIMUM_SPEED) && \
```
Comment (jfischer-phytec-iot) [inline include/usb/usb_device.h]: I think doing it this way is not good, readability suffers and worse, we need to touch it again to support more then one device controller, e.g. a SoC may have one FS controller and one HS controller. That why I want to check it [explicit ](https://github.com/zephyrproject-rtos/zephyr/pull/15147/commits/9657a10e2bedf655f5b27445ef39c8147e0442c8)for at least HS and FS.

**Review by jfischer-no — DISMISSED** [review]

File: include/usb/usb_device.h (hunk 2 above)
Comment (finikorg) [reply include/usb/usb_device.h]: To support several device controllers we need to rewrite a lot anyway, so it is probably better to touch less code. And we do compare against one global `DT_USBHS_MAXIMUM_SPEED` which does not depends on the controller. Probably that check should be done in `usb_dc_***` function.

Comment (finikorg) [reply include/usb/usb_device.h]: 3c750df3ddf6ba464389b49f69273a35abfbb118 make usb device controller check

**Comment by carlescufi** [comment]
Will be reworked

**Comment by carlescufi** [comment]
@jfischer-phytec-iot can you re-review this one?

**Comment by galak** [comment]
@finikorg is this still relevant?  If not please close or update the PR.

**Comment by finikorg** [comment]
Closing for now

**Pull request closed without merging**